src/shogun/base/SGObject.h

```diff
@@ -795,6 +796,64 @@ class CSGObject
 	}
 
 protected:
+
+	template <
+			ParameterProperties PPVal, typename... Args,
+			std::enable_if_t<
+					static_cast<bool>(PPVal& ParameterProperties::AUTO)>* = nullptr>
+	void declare_auto_func(AnyParameter& any_pprop, Args&&... args)
+	{
+		static_assert(
+				sg_is_any_base_of<params::AutoInit, Args...>::value,
+				"Expected a params::AutoInit function when passing param with"
+				" ParameterProperty::AUTO!");
+		constexpr size_t idx =
+				get_idx_from_pack<params::AutoInit, Args...>::idx;
+		any_pprop.set_init_function(
+				std::move(std::get<idx>(std::forward_as_tuple(args...))));
+	}
+
+	template <
+			ParameterProperties PPVal, typename ValueType, typename... Args,
+			std::enable_if_t<static_cast<bool>(
+					PPVal& ParameterProperties::CONSTRAIN)>* = nullptr>
+	void declare_constraint_func(AnyParameter& any_pprop, Args&&... args)
+	{
+		static_assert(
+				sg_is_any_base_of<ConstraintBase, Args...>::value,
+				"Expected a Constraint object when passing param with"
+				" ParameterProperty::CONSTRAIN!");
+		constexpr size_t idx =
+				get_idx_from_pack<ConstraintBase, Args...>::idx;
+		auto func = [constrain_function = std::move(
+				std::get<idx>(std::forward_as_tuple(args...)))](
+				const auto& val) {
+			std::string result;
+			auto casted_val = any_cast<ValueType>(val);
+			constrain_function.run(casted_val, result);
+			return result;
+		};
+		any_pprop.set_constrain_function(std::move(func));
+	}
+
+	template <ParameterProperties pprop, typename T, typename... Args>
+	void declare(
+			T* value, const std::string& name, const std::string& description,
+			Args&&... args)
+	{
+		auto any_pprop = AnyParameterProperties(description, pprop);
+		auto anyp = AnyParameter(make_any_ref(value), any_pprop);
+		if constexpr (static_cast<bool>(pprop & ParameterProperties::AUTO))
+			declare_auto_func<pprop>(anyp, std::forward<Args&&>(args)...);
+		if constexpr (static_cast<bool>(
+				pprop & ParameterProperties::CONSTRAIN))
+			declare_constraint_func<pprop, T>(
+					anyp, std::forward<Args&&>(args)...);
```

Comment:
so these two calls actually work well together, and now we can add a constraint and auto initialiser in whichever order, to the same parameter. We can keep on adding more callbacks with parameter registration this way if we want in a type safe way

Reply:
very nice! :)

Reply:
btw I have been experimenting with making AnyParameter constexpr, but the default parameter mask that I added in the other PR won't allow it. long story short the default mask is non constant evaluated unless we make it a non type template, which would make SGObject templated.. so that is a no go

Reply:
This would have been neat:

```cpp
template <ParameterProperties pprop = ParameterProperties::NONE, typename T, typename ConstexprLambda, typename... Args>
void declare(T* value, const std::string& name, ConstexprLambda&& description, Args&&... args)
{
//	    constexpr auto mask = pprop | m_const_default_mask; <- breaks everything
	constexpr auto any_pprop = AnyParameterProperties(description(), pprop);
	auto anyp = AnyParameter(make_any_ref(value), any_pprop);
	if constexpr (any_pprop.has_property(ParameterProperties::AUTO))
		declare_auto_func<pprop>(anyp, std::forward<Args&&>(args)...);
	if constexpr (any_pprop.has_property(ParameterProperties::CONSTRAIN))
		declare_constraint_func<pprop, T>(
				anyp, std::forward<Args&&>(args)...);
	BaseTag tag(name);
	create_parameter(tag, anyp);
}
```
However only in C++20 will we be able to pass string literals as non type templates, that's why the description argument is actually a lambda that returns the description (lambdas are const evaluated in C++17). It makes the code a bit ugly:
```cpp
#define MAKE_CONSTEXPR_LAMBDA(x) [](){return x;}

declare<ParameterProperties::HYPER>(&coef0, "coef0", MAKE_CONSTEXPR_LAMBDA("Coefficient 0."));
```
Anyway, I guess this is a no go for now...

Reply:
> btw I have been experimenting with making AnyParameterProperties constexpr, but the default parameter mask that I added in the other PR won't allow it. long story short the default mask is non constant evaluated unless we make it a non type template, which would make SGObject templated.. so that is a no go

I think the default mask is something less important. Maybe we do this differently in favour of having the map itself compile time?

Reply:
oh this is unrelated to the map being compile time. I am also working on that now, just getting it to complain when a key already exists. The thing I am talking about her is to make AnyParameterProperties compile time because all the information is there at compile time, but the performance gain of this is minimal..

Reply:
Is constexpr and end in itself? :D

Reply:
well it's just annoying to have to execute code at runtime that is known at compile time :p